rust-runtime/aws-smithy-runtime/src/client/connectors/connection_poisoning.rs

```diff
@@ -0,0 +1,132 @@
+    BeforeDeserializationInterceptorContextMut, BeforeTransmitInterceptorContextMut,
+};
+use aws_smithy_runtime_api::client::interceptors::Interceptor;
+use aws_smithy_runtime_api::client::retries::{ClassifyRetry, RetryReason};
+use aws_smithy_types::config_bag::{ConfigBag, Storable, StoreReplace};
+use aws_smithy_types::retry::{ErrorKind, ReconnectMode, RetryConfig};
+use std::fmt;
+use tracing::{debug, error};
+
+/// A interceptor for poisoning connections in response to certain events.
+///
+/// This interceptor, when paired with a compatible connection, allows the connection to be
+/// poisoned in reaction to certain events *(like receiving a transient error.)* This allows users
+/// to avoid sending requests to a server that isn't responding. This can increase the load on a
+/// server, because more connections will be made overall.
+///
+/// **In order for this interceptor to work,** the configured connection must interact with the
+/// "connection retriever" stored in an HTTP request's `extensions` map. For an example of this,
+/// see [aws_smithy_client::hyper_ext::Adapter](https://github.com/awslabs/smithy-rs/blob/47b3d23ff3cabd67e797af616101f5a4ea6be5e8/rust-runtime/aws-smithy-client/src/hyper_ext.rs#L155).
+/// When a connection is made available to the retriever, this interceptor will call a `.poison`
+/// method on it, signalling that the connection should be dropped. It is up to the connection
+/// implementer to handle this.
+#[non_exhaustive]
+#[derive(Debug, Default)]
+pub struct ConnectionPoisoningInterceptor {}
+
+impl ConnectionPoisoningInterceptor {
+    pub fn new() -> Self {
+        Self::default()
+    }
+}
+
+impl Interceptor for ConnectionPoisoningInterceptor {
+    fn modify_before_transmit(
+        &self,
+        context: &mut BeforeTransmitInterceptorContextMut<'_>,
+        cfg: &mut ConfigBag,
+    ) -> Result<(), BoxError> {
+        let capture_smithy_connection = CaptureSmithyConnectionWrapper::new();
+        context
+            .request_mut()
+            .extensions_mut()
+            .insert(capture_smithy_connection.clone_inner());
+        cfg.interceptor_state().store_put(capture_smithy_connection);
+
+        Ok(())
+    }
+
+    fn modify_before_deserialization(
+        &self,
+        context: &mut BeforeDeserializationInterceptorContextMut<'_>,
+        cfg: &mut ConfigBag,
+    ) -> Result<(), BoxError> {
+        let reconnect_mode = cfg
+            .load::<RetryConfig>()
+            .map(RetryConfig::reconnect_mode)
+            .unwrap_or(ReconnectMode::ReconnectOnTransientError);
+        let captured_connection = cfg.load::<CaptureSmithyConnectionWrapper>().cloned();
+        let retry_classifiers = cfg.retry_classifiers();
+
+        let error_is_transient = retry_classifiers
+            .classify_retry(context.into_inner())
+            .map(|reason| reason == RetryReason::Error(ErrorKind::TransientError))
+            .unwrap_or_default();
```

Comment:
Maybe the orchestrator should add the retry classification to the ConfigBag?

Reply:
Yeah, I can see how that could be useful to users. However, the retry strategy doesn't have write access to the bag, and it's the thing that's calling the error classifier (not the orchestrator.)

I'm going to leave this as-is for now.